src/scalarstats.jl

```diff
@@ -261,36 +261,92 @@ realXcY(x::Real, y::Real) = x*y
-    y = iterate(x)
-    if y === nothing
-        T = eltype(x)
-        # Return the NaN of the type that we would get, had this collection
-        # contained any elements (this is consistent with std)
-        return oftype(sqrt((abs2(zero(T)) + abs2(zero(T)))/2), NaN)
+    sem(x; mean=nothing)
+    sem(x::AbstractArray[, weights::AbstractWeights]; mean=nothing)
+
+Return the standard error of the mean for a collection `x`. When using no weights, this is
+equal to the (sample) standard deviation divided by the sample size. If weights are used, 
+the variance of the sample mean is calculated as follows:
+
+* `AnalyticWeights`: Not implemented.
+* `FrequencyWeights`: ``\\frac{\\sum_{i=1}^n w_i (x_i - \\bar{x_i})^2}{(\\sum w_i) (\\sum w_i - 1)}``
+* `ProbabilityWeights`: ``\\frac{n}{n-1} \\frac{\\sum_{i=1}^n w_i^2 (x_i - \\bar{x_i})^2}{\\left( \\sum w_i \\right)^2}``
+
+The standard error is then the square root of the above quantities.
+
+# References
+
+Carl-Erik Särndal, Bengt Swensson, Jan Wretman (1992). Model Assisted Survey Sampling.
+New York: Springer. pp. 51-53.
+"""
+sem(x, weights::AbstractWeights; mean=nothing) = _sem(x, weights, mean)
+sem(x; mean=nothing) = _sem(x, mean)
+
+function _sem(x, mean)
+    n = 0
+    if isempty(x)
+        # Return the NaN of the type that we would get for a nonempty x
+        variance = var(x; mean=mean, corrected=true)
+    else
+        sse = real(zero(mean))
+        for element in x
+            n += 1
+            sse += abs2(element - mean)
+        end
+        variance = sse / (n - 1)
     end
-    count = 1
-    value, state = y
-    y = iterate(x, state)
-    # Use Welford algorithm as seen in (among other places)
-    # Knuth's TAOCP, Vol 2, page 232, 3rd edition.
-    M = value / 1
-    S = real(zero(M))
-    while y !== nothing
+    return sqrt(variance / n)
+end
+
+function _sem(x, ::Nothing)
+    n = 0
+    if isempty(x)
+        # Return the NaN of the type that we would get for a nonempty x
+        variance = var(x; corrected=true)
+    else
+        y = iterate(x)
         value, state = y
-        y = iterate(x, state)
-        count += 1
-        new_M = M + (value - M) / count
-        S = S + realXcY(value - M, value - new_M)
-        M = new_M
+        # Use Welford algorithm as seen in (among other places)
+        # Knuth's TAOCP, Vol 2, page 232, 3rd edition.
+        mean = value / 1
+        sse = real(zero(mean))
+        while y !== nothing
+            value, state = y
+            y = iterate(x, state)
+            n += 1
+            new_mean = mean + (value - mean) / n
+            sse += realXcY(value - mean, value - new_mean)
+            mean = new_mean
+        end
+        variance = sse / (n - 1)
     end
-    var = S / (count - 1)
-    return sqrt(var/count)
+    return sqrt(variance / n)
+end
+
+_sem(x::AbstractArray, mean) = sqrt(var(x; mean=mean, corrected=true) / length(x))
+_sem(x::AbstractArray, ::Nothing) = sqrt(var(x; corrected=true) / length(x))
```

Comment:
No need for two basically identical methods:.
```suggestion
sem(x::AbstractArray; mean=nothing) = sqrt(var(x; mean=mean, corrected=true) / length(x))
```

Reply:
Without it there's a method ambiguity.

Reply:
Are you sure? Note that the suggestion is to implement `sem`, not `_sem`.

Reply:
Ahh, I see.